### DISCUSSION TITLE
Using brax to learn robot dynamics

### DISCUSSION THREAD

**Discussion by shamilmamedov** [google/brax#291] [Q&A] [answered]
Hi there! I have got a few general questions about brax, I would appreciate if experienced users of brax can clarify them for me.

- Is brax the right tool for learning the dynamics of a robot consisting of rigid bodies connected by multi-dof elastic joints? 
- Is there a way to get the ODE of the system in the form `f(theta, x, u)` where `theta` is the vector of parameters? 
- Is there a way to specify arbitrary joint force, e.g., use a neural network as a spring force?

**Comment by btaba** [comment, marked as answer]
Hi @shamilmamedov , below is an attempt to answer some of the questions (hope it's helpful):
* Brax can help learn dynamics of an articulated rigid body robot with multi-dof joints. We don't currently model muscles or tendons (if that's what you mean by elastic)
* The ODE depends on the backend used, we have multiple backends in brax. `f` as you describe it, would be `pipeline.step` [here](https://github.com/google/brax/blob/main/brax/v2/spring/pipeline.py#L54) or [here](https://github.com/google/brax/blob/main/brax/v2/generalized/pipeline.py#L54). For some background on the generalized physics backend, see chapter 4 [here](https://people.eecs.berkeley.edu/~jfc/mirtich/thesis/mirtichThesis.pdf), or for spring see [here](https://danielchappuis.ch/download/ConstraintsDerivationRigidBody3D.pdf). I'm sure there are better references but those come top of mind
* To specify an arbitrary joint force, you can use an actuator. When you call `pipeline.step(sys, state, action)` you can plug in the `action` from your neural net
